### PR TITLE
check the shape for mat, csr and csc in prediction

### PR DESCRIPTION
**Pull request by guolinke**
partially improve #812

### CODE REVIEW

**Comment by StrikerRUS** [comment]
Close-reopen for CI.

File: src/c_api.cpp [hunk 1]
```diff
@@ -249,17 +249,19 @@ class Booster {
     boosting_->RollbackOneIter();
   }
 
-  void PredictSingleRow(int num_iteration, int predict_type,
+  void PredictSingleRow(int num_iteration, int predict_type, int ncol,
                std::function<std::vector<std::pair<int, double>>(int row_idx)> get_row_fun,
                const Config& config,
                double* out_result, int64_t* out_len) {
+    if (ncol != boosting_->MaxFeatureIdx() + 1) {
+      Log::Fatal("The number of feature in data (%d) is not same as in training (%d).", ncol, boosting_->MaxFeatureIdx() + 1);
```
Comment (StrikerRUS) [inline src/c_api.cpp]: Maybe better
```
"The number of features in data (%d) is not the same as it was in training data (%d)."
```
?

File: src/c_api.cpp [hunk 2]
```diff
@@ -268,10 +270,13 @@ class Booster {
   }
 
 
-  void Predict(int num_iteration, int predict_type, int nrow,
+  void Predict(int num_iteration, int predict_type, int nrow, int ncol,
                std::function<std::vector<std::pair<int, double>>(int row_idx)> get_row_fun,
                const Config& config,
                double* out_result, int64_t* out_len) {
+    if (ncol != boosting_->MaxFeatureIdx() + 1) {
+      Log::Fatal("The number of feature in data (%d) is not same as in training (%d).", ncol, boosting_->MaxFeatureIdx() + 1);
```
Comment (StrikerRUS) [inline src/c_api.cpp]: The same as above.

File: src/c_api.cpp [hunk 3]
```diff
@@ -1315,7 +1320,7 @@ int LGBM_BoosterPredictForCSR(BoosterHandle handle,
   Booster* ref_booster = reinterpret_cast<Booster*>(handle);
   auto get_row_fun = RowFunctionFromCSR(indptr, indptr_type, indices, data, data_type, nindptr, nelem);
   int nrow = static_cast<int>(nindptr - 1);
-  ref_booster->Predict(num_iteration, predict_type, nrow, get_row_fun,
+  ref_booster->Predict(num_iteration, predict_type, nrow, static_cast<int32_t>(num_col), get_row_fun,
```
Comment (StrikerRUS) [inline src/c_api.cpp]: Is it safe? Maybe simply change the original type, as it was not used before?

File: src/c_api.cpp [hunk 4]
```diff
@@ -1343,7 +1348,7 @@ int LGBM_BoosterPredictForCSRSingleRow(BoosterHandle handle,
   }
   Booster* ref_booster = reinterpret_cast<Booster*>(handle);
   auto get_row_fun = RowFunctionFromCSR(indptr, indptr_type, indices, data, data_type, nindptr, nelem);
-  ref_booster->PredictSingleRow(num_iteration, predict_type, get_row_fun, config, out_result, out_len);
+  ref_booster->PredictSingleRow(num_iteration, predict_type, static_cast<int32_t>(num_col), get_row_fun, config, out_result, out_len);
```
Comment (StrikerRUS) [inline src/c_api.cpp]: The same as above.

File: src/c_api.cpp [hunk 5]
```diff
@@ -268,10 +270,13 @@ class Booster {
   }
 
 
-  void Predict(int num_iteration, int predict_type, int nrow,
+  void Predict(int num_iteration, int predict_type, int nrow, int ncol,
                std::function<std::vector<std::pair<int, double>>(int row_idx)> get_row_fun,
                const Config& config,
                double* out_result, int64_t* out_len) {
+    if (ncol != boosting_->MaxFeatureIdx() + 1) {
```
Comment (StrikerRUS) [inline src/c_api.cpp]: Doesn't it break the part about `0`?
https://github.com/microsoft/LightGBM/blob/1f88721a7982a804731b1dc108ee40625405639c/include/LightGBM/c_api.h#L686

**Comment by StrikerRUS** [comment]
Please add a simple test from the original issue:
```
import numpy as np
import lightgbm as lgb

x_data = np.random.rand(100, 10)
x_bad_data = np.random.rand(100, 11)
y_data =  np.random.rand(100) > .5
self.assertNotEqual(x_data.shape[-1], x_bad_data.shape[-1])
train_dataset = lgb.Dataset(x_data, y_data)
bst = lgb.train({'objective': 'binary'}, train_dataset)
with np.testing.assert_raises_regex(lgb.basic.LightGBMError,
                                    'The number of features in data*'):
    bst.predict(x_bad_data)
```

File: src/c_api.cpp (hunk 5 above)
Comment (guolinke) [reply src/c_api.cpp]: just realize we actually didn't guss the num_col when it is zero. 
I think it is better to pass non-zero values.

**Comment by guolinke** [comment]
There are more changes than I expected. So I think we should test all cases, including the the mat, libsvm file and CSR format. 
@StrikerRUS could you help for the test cases?

**Comment by StrikerRUS** [comment]
@guolinke 
> could you help for the test cases?

Sure!

What do you think about changing the original type to avoid casting? https://github.com/microsoft/LightGBM/pull/2464#discussion_r329257560

Does this PR already include [that check](https://github.com/microsoft/LightGBM/issues/2457#issuecomment-535911772)?

File: src/c_api.cpp (hunk 3 above)
Comment (guolinke) [reply src/c_api.cpp]: change the original will cause lots of API changes. 
I think we actually don't support too many columns.
I will add a check for the num_col in CSR format, to ensure its value is smaller than int32_max.

**Comment by guolinke** [comment]
> Does this PR already include that check?
Yeah, as the zero-based and one-based libsvm format have the different number of columns.

**Comment by StrikerRUS** [comment]
@guolinke I added tests in the latest commit.

As we do not support a case when data for prediction is a list of arrays, I tried to check it for validation. And it seems that validation shape is not covered by this PR, right?


```
bad_valid_data = train_data.create_valid(bad_X_test, label=y_test)
bst.add_valid(bad_valid_data, "valid_bad")
bst.eval_valid()
```

**Comment by guolinke** [comment]
yeah, validation is not converted for now.

**Comment by guolinke** [comment]
due the need of pr #2485, I will merge this

**Pull request merged**